File: cses/1634_minimizing_coins.cpp
#include <bits/stdc++.h>
using namespace std;

typedef unordered_map<int, unordered_map<int, int>> Cache;
typedef unordered_map<int, int> Ints;

int solution(int target, vector<int>* values, int k, Cache* cache) {
  Cache::iterator cit;
  // cout << target << " -- " << k << endl;
  // for(cit = cache->begin(); cit != cache->end(); ++cit){
  //   cout << cit->first << " " << cit->second.size() << endl;
  // }
  if(target == 0){
    return 0;
  }
  // if((cit = cache->find(target)) != cache->end()){
  //   Ints::iterator it;
  //   if((it = cit->second.find(k)) != cit->second.end()){
  //     return it->second;
  //   }
  // }
  int res = 1e9;
  int value = (*values)[k];
  int q = target / value;
  int r = target % value;
  int s = 0;

  if(r == 0){
    res = q;
  } else if(k != 0){
    for(;q >= 0; --q){
      s = solution(r, values, k - 1, cache);
      res = res < q + s ? res : q + s;
      r += value;
    }
  }
  // if((cit = cache->find(target)) == cache->end()){
  //   unordered_map<int, int> mm;
  //   mm[k] = res;
  //   (*cache)[target] = mm;
  // }else{
  //   cit->second[k] = res;
  // }
  return res;
}

int main() {
  int n, target;
  cin >> n >> target;
  vector<int> c(n);
  vector<int> values;
  for (int&v : c) cin >> v;

  // vector<int> dp(target+1,1e9);
  // dp[0] = 0;
  // for (int i = 1; i <= target; i++) {
  //   for (int j = 0; j < n; j++) {
  //     if (i-c[j] >= 0) {
	// dp[i] = min(dp[i], dp[i-c[j]]+1);
  //     }
  //   }
  // }
  // cout << (dp[target] == 1e9 ? -1 : dp[target]) << endl;
  Cache cache;
  sort(c.begin(), c.end());
  vector<int>::iterator it = unique(c.begin(), c.begin() + c.size());
  for(it = c.begin(); it != c.end(); ++it){
    values.push_back(*it);
  }
  int res = solution(target, &values, values.size() - 1, &cache);
  cout << res << endl;
}

File: cses/1634_minimizing_coins.py
from functools import lru_cache

def solution(n, values, target):
    M = target + 1
    # values = set(values)
    dp = [M for i in range(target + 1)]
    dp[0] = 0
    for i in range(1, target + 1):
        for v in values:
            if i - v >= 0:
                dp[i] = min(dp[i], 1 + dp[i - v])
    return dp[-1] if dp[-1] < M else -1


def solution2(n, values, target):
    M = target + 1
    
    @lru_cache(None)
    def sol(t, k):
        if t == 0:
            return 0
        if k < 0:
            return M
        res = M
        value = values[k]
        q, r = divmod(t, value)
        if r == 0:
            return q
        if k == 0:
            return M
        for i in range(q, -1, -1):
            res = min(res, i + sol(r, k - 1))
            r += value
        return res

    values = set(values)
    values = sorted(values)
    # print(len(values))
    res = sol(target, len(values) - 1)
    return res if res < M else -1



def main():
    n, target = list(map(int, input().split()))
    values = list(map(int, input().split()))
    res = solution2(n, values, target)
    print(res) 

main()